relay-server/src/actors/global_config.rs

```diff
@@ -253,25 +251,57 @@ impl Service for GlobalConfigService {
             let mut shutdown_handle = Controller::shutdown_handle();
 
+            let config = &self.config;
             relay_log::info!("global config service starting");
-
-            if self.config.has_credentials() {
-                // NOTE(iker): if this first request fails it's possible the default
-                // global config is forwarded. This is not ideal, but we accept it
-                // for now.
-                self.update_global_config();
-            } else {
-                // NOTE(iker): not making a request results in the sleep handler
-                // not being reset, so no new requests are made.
-                relay_log::info!("fetching global configs disabled: no credentials configured");
-            }
+            match (
+                config.relay_mode(),
+                config.has_credentials(),
+                config.global_config(),
+            ) {
+                (RelayMode::Managed, true, global_config) => {
+                    if global_config.is_some() {
+                        relay_log::info!("ignoring static global config in managed mode");
+                    }
+                    relay_log::info!("serving global configs fetched from upstream");
+                    // This request will trigger the request intervals when internal_rx receives the
+                    // result from upstream.
+                    self.request_global_config();
+                }
+                (RelayMode::Managed, false, _) => {
+                    // NOTE(iker): not making a request results in the sleep handler
+                    // not being reset, so no new requests are made.
+                    relay_log::info!("serving default global configs due to lacking credentials");
+                }
+                (
+                    RelayMode::Proxy | RelayMode::Static | RelayMode::Capture,
+                    _,
+                    Some(global_config),
+                ) => {
+                    relay_log::info!("serving static global config loaded from file");
+                    self.global_config_watch.send(global_config.clone()).ok();
+                }
+                (RelayMode::Proxy | RelayMode::Static | RelayMode::Capture, _, None) => {
+                    relay_log::info!(
+                        "serving default global configs due to lacking static global config file"
+                    );
+                }
+            };
 
             loop {
                 tokio::select! {
                     biased;
+                    () = &mut self.fetch_handle => {
+                        // Disable upstream requests timer until we receive result of query.
+                        self.fetch_handle.reset();
+
+                        self.request_global_config();
+                    }
+                    Some(result) = self.internal_rx.recv() => {
+                        // Enable upstream requests timer for global configs.
+                        self.schedule_fetch();
```

Comment:
What's the motivation to move the `reset` and `schedule_fetch` from the functions to this `select!`? I see no added benefit and find it harder to read. I'd suggest not modifying this code unless it's really needed.

Also, the new schedule should happen after handling the result from upstream.

Reply:
Personally I find it easier to read, as it seems more like a side-effect that you don't expect when reading from the caller. However, if it's indeed harder to read for other people then I'll revert it

Reply:
I've reverted it now